proposals/simd-0170-builtin-instruction-cost-and-budget.md

```diff
@@ -0,0 +1,99 @@
+---
+simd: '0170'
+title: Reserve minimal CUs for builtins
+authors:
+  - Tao Zhu (Anza)
+category: Standard
+type: Core
+status: Review
+created: 2024-08-26
+feature: https://github.com/anza-xyz/agave/issues/2562
+supersedes: 
+superseded-by:
+extends:
+---
+
+## Summary
+
+If a transaction doesn't request a compute budget limit, then for each builtin
+program instruction allocate 3,000 compute units rather than 200,000.
+
+## Motivation
+
+This proposal addresses an issue related to CU allocation and consumption
+for builtin programs while aiming to avoid adding unnecessary complexity.
+
+- Preventing over-allocation of CUs: Over-allocating CUs for builtins
+   reduces block density, lowers network throughput, and can degrade block
+producer performance by causing repeated transaction retries. Avoiding excessive
+CU allocation is critical for maximizing block efficiency and minimizing network
+delays.
+
+To resolve this issue, we propose statically defining the CU reservation of
+builtin instructions based on worst-case scenarios, including potential CPI
+calls. A unified MAX_BUILTIN_ALLOCATION_COMPUTE_UNIT_LIMIT will standardize CU
+allocation for both block producers and the CU meter.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+Establish a static maximum CU allocation: Define a global compute unit limit
+of 3,000 CUs, denoted as MAX_BUILTIN_ALLOCATION_COMPUTE_UNIT_LIMIT, to
+accommodate worst-case execution scenarios, including potential Cross-Program
+Invocations (CPIs). This uniform limit is applied to each builtin program
+instruction and is used to configure both CU meter allocations and block
+producer CU limits reservation for all builtin instructions.
+
```

Comment:
The static list of builtin program id's that will have 3000 compute units allocated are:
- `Stake11111111111111111111111111111111111111`
- `Config1111111111111111111111111111111111111`
- `Vote111111111111111111111111111111111111111`
- `11111111111111111111111111111111`
- `ComputeBudget111111111111111111111111111111`
- `AddressLookupTab1e1111111111111111111111111`
- `BPFLoaderUpgradeab1e11111111111111111111111`
- `BPFLoader1111111111111111111111111111111111`
- `BPFLoader2111111111111111111111111111111111`
- `LoaderV411111111111111111111111111111111111`
- `KeccakSecp256k11111111111111111111111111111`
- `Ed25519SigVerify111111111111111111111111111`

Note that there are a few builtin programs not in the list including the upcoming zk programs as well as the feature program.